src/pydata_sphinx_theme/assets/styles/extensions/_notebooks.scss

```diff
@@ -9,84 +9,202 @@
+    div.prompt pre::before {
+      content: "In ";
+    }
+
+    div.input_area {
+      border: none; // border taken care of by <pre> styles elsewhere in the theme
+      overflow: visible;
+
+      > div.highlight {
+        overflow: visible; // necessary for focus ring
+      }
+    }
+  }
 
-    // Dark theme special-cases
-    html[data-theme="dark"] & {
-      &.rendered_html:not(:has(table.dataframe)),
-      // ipywidgets
-      .widget-subarea {
-        @include cell-output-background;
+  // Output-specific styles
+  div.nboutput.container {
+    // Space between the input portion and the output portion. Cannot use margin
+    // here because we do not want to create a break in the vertical bar running
+    // down the left side.
+    padding-top: $top-and-bottom-container-space;
+
+    // Prefix the cell number with the word "Out". TODO: add to translations
+    div.prompt pre::before {
+      content: "Out ";
+    }
+
+    div.output_area {
+      &.rendered_html,
+      div[class*="highlight"].math,
+      > div.output_javascript,
+      > img {
+        // Override nbsphinx. We want the left edge of the output area to be
+        // left-aligned with the output label, the input area, and the input
+        // label.
+        padding: 0;
       }
 
+      // Override nbsphinx. Things like box shadows on elements that are
+      // children of div.output_area do not work if the overflow is not set to
+      // visible. But do not apply to .rendered_html because it could be, for
+      // example, a long pandas data table that is rendered without a
+      // non-overflowing container.
+      &:not(.rendered_html) {
+        overflow: visible;
+      }
+
+      // What this selector attempts to do is to override a bunch of CSS rules
+      // that come from multiple places that add top and bottom whitespace to elements,
+      // like paragraph elements. These rules mess up the alignment between
+      // bottom of the notebook cell content and the left bar that we've added.
+      // They also create unnecessary extra whitespace at the bottom of notebook
+      // cells. However, this selector may be a bit too wide, too blunt. We may
+      // need find a different way to achieve this.
+      :not(.highlight, .widget-tab-contents, .jupyter-widgets):last-child {
+        margin-top: 0;
+        margin-bottom: 0;
+      }
+
+      div.highlight,
+      div.nohighlight,
+      .jupyter-widgets,
+      .jupyter-matplotlib-canvas-container {
+        overflow: visible; // for focus ring
+      }
+
+      // use theme color for stderr
       &.stderr {
-        background-color: var(--pst-color-danger);
+        background-color: transparent;
+
+        div[class*="highlight"] > pre {
+          background-color: var(--pst-color-danger-bg);
+        }
+      }
+
+      // Override our own style rule in _math.scss
+      div.math {
+        display: block;
+      }
+
+      span.math {
+        display: inline;
```

Comment:
I will open a follow-up PR to better handle math content both within notebooks and outside of notebooks. That way we won't need to override our own rules.